interfaces/builtin/custom_device.go

```diff
@@ -0,0 +1,394 @@
+	snippet := &bytes.Buffer{}
+	emitRule := func(paths []string, permissions string) {
+		for _, path := range paths {
+			fmt.Fprintf(snippet, "\"%s\" %s,\n", path, permissions)
+		}
+	}
+
+	// get all attributes without validation, since that was done before;
+	// should an error occur, we'll simply not write any rule.
+
+	var devicePaths []string
+	_ = slot.Attr("devices", &devicePaths)
+	emitRule(devicePaths, "rw")
+
+	var readDevicePaths []string
+	_ = slot.Attr("read-devices", &readDevicePaths)
+	emitRule(readDevicePaths, "r")
+
+	var writablePaths []string
+	_ = slot.Attr("write", &writablePaths)
+	emitRule(writablePaths, "rw")
+
+	var readOnlyPaths []string
+	_ = slot.Attr("read", &readOnlyPaths)
+	emitRule(readOnlyPaths, "r")
+
+	spec.AddSnippet(snippet.String())
+	return nil
+}
+
+// extractStringMapAttribute looks up the given key in the container, and
+// returns its value as a map[string]string.
+// No validation is performed, since it already occurred before connecting the
+// interface.
+func (iface *customDeviceInterface) extractStringMapAttribute(container map[string]interface{}, key string) map[string]string {
+	valueMap, ok := container[key].(map[string]interface{})
+	if !ok {
+		return nil
+	}
+
+	stringMap := make(map[string]string, len(valueMap))
+	for key, value := range valueMap {
+		stringMap[key] = value.(string)
+	}
+
+	return stringMap
+}
+
+func (iface *customDeviceInterface) UDevConnectedPlug(spec *udev.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	// Collect all the device paths specified in either the "devices" or
+	// "read-devices" attributes.
+	var devicePaths []string
+	_ = slot.Attr("devices", &devicePaths)
+	var readDevicePaths []string
+	_ = slot.Attr("read-devices", &readDevicePaths)
+	allDevicePaths := devicePaths
+	allDevicePaths = append(allDevicePaths, readDevicePaths...)
+
+	// Generate a basic udev rule for each device; we put them into a map
+	// indexed by the device name, so that we can overwrite the entry later
+	// with a more specific rule.
+	deviceRules := make(map[string]string, len(allDevicePaths))
+	for _, devicePath := range allDevicePaths {
+		if strings.HasPrefix(devicePath, "/dev/") {
+			deviceName := devicePath[5:]
+			deviceRule := fmt.Sprintf(`KERNEL=="%s"`, deviceName)
+			deviceRules[deviceName] = deviceRule
+		}
+	}
+
+	// Generate udev rules from the "udev-tagging" attribute; note that these
+	// rules might override the simpler KERNEL=="<device>" rules we computed
+	// above -- that's fine.
+	var udevTaggingRules []map[string]interface{}
+	_ = slot.Attr("udev-tagging", &udevTaggingRules)
+	for _, udevTaggingRule := range udevTaggingRules {
+		rule := &bytes.Buffer{}
+
+		deviceName, ok := udevTaggingRule["kernel"].(string)
+		if !ok {
+			return customDeviceInternalError
+		}
+
+		fmt.Fprintf(rule, `KERNEL=="%s"`, deviceName)
```

Comment:
I prefer to keep it like this, because I find the rule more recognizeable in this way. Also, I doubt that golang will ever change the quoting or escaping rules for their strings, but I feel it's safer if we don't let golang do any processing of the string here (just to be 100% sure that our behaviour will not change across versions of the golang compiler).